pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,297 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// reconciler is responsible for transforming current EndpointSlice state into
+// desired state
+type reconciler struct {
+	client               clientset.Interface
+	nodeLister           corelisters.NodeLister
+	maxEndpointsPerSlice int32
+}
+
+// endpointMeta includes the attributes we group slices on, this type helps with
+// that logic in reconciler
+type endpointMeta struct {
+	Ports       []discovery.EndpointPort `json:"ports" protobuf:"bytes,2,rep,name=ports"`
+	AddressType *discovery.AddressType   `json:"addressType" protobuf:"bytes,3,rep,name=addressType"`
+}
+
+// reconcile takes a set of pods currently matching a service selector and
+// compares them with the endpoints already present in any existing endpoint
+// slices for the given service. It creates, updates, or deletes endpoint slices
+// to ensure the desired set of pods are represented by endpoint slices.
+func (r *reconciler) reconcile(service *corev1.Service, pods []*corev1.Pod, existingSlices []*discovery.EndpointSlice, triggerTime time.Time) error {
```

Comment:
The below will work.

My concern is `fragmentation` and `lack of compaction`. Do we want to favor `large # of slices with small # of endpoints` or `small # of slices with large # of endpoints`? in a scenario where the # of pods fronted by the same service grows multiple orders of magnitude then back down the below will leave behind many endpoint slices with very little utilization each.

Reply:
We want to have `small # of slices with large # of endpoints. 

Yes. What you say can happen since the algorithm does not have active rebalancing. But one rolling update of the backends will automatically repack the endpoints.

Reply:
I am not sure how a rolling update will rebalance the slices. However the time between scale-down and a rolling update can be too long to carry forward with fragmented slices.

Reply:
The working theory is that by always adding new endpoints to slices that will end up closest to full, the smallest slices will eventually drop off naturally, especially if there's a rolling update where no old endpoint will be left behind. Definitely plan on some more thorough testing of this algorithm at scale, and likely some adjustments. The main goal for this alpha was to have something simple and reliable that was reasonably performant.